In [1]:
import pandas as pd
import numpy as np

In [3]:
#reading csv file and creating a dataframe
df = pd.read_csv("E:\project\Chickpea.data.csv")
df.head()

Predictor       740       741       742       743       744       745  \
0     Dimtu  1.312229  1.313752  1.315363  1.317040  1.318748  1.320468   
1     Dimtu  1.207945  1.209409  1.211038  1.212806  1.214665  1.216582   
2     Dimtu  1.382931  1.383910  1.385126  1.386556  1.388142  1.389842   
3     Dimtu  1.464557  1.465710  1.466986  1.468363  1.469794  1.471253   
4     Dimtu  1.145023  1.146660  1.148481  1.150452  1.152523  1.154657   

        746       747       748    ...         1061      1062      1063  \
0  1.322220  1.324037  1.325907    ...     1.539160  1.540128  1.539360   
1  1.218553  1.220593  1.222671    ...     1.491851  1.492835  1.492130   
2  1.391645  1.393558  1.395535    ...     1.566846  1.567866  1.567114   
3  1.472753  1.474325  1.475947    ...     1.646721  1.647711  1.646850   
4  1.156848  1.159107  1.161403    ...     1.454652  1.455651  1.454998   

       1064      1065      1066      1067      1068      1069      1070  
0  1.537472  1.535576  1.533732  1.534212  1.536558  1.539456  1.541060  
1  1.490335  1.488528  1.486766  1.487255  1.489549  1.492376  1.493944  
2  1.565218  1.563311  1.561454  1.561961  1.564364  1.567328  1.568971  
3  1.644794  1.642736  1.640737  1.641228  1.643717  1.646801  1.648502  
4  1.453276  1.451538  1.449841  1.450336  1.452587  1.455357  1.456896  

[5 rows x 332 columns]

In [4]:
#Checking NA values
df.isnull().values.any()

False

In [6]:
#collecting independent variable to X
X = df.drop(['Predictor'], axis=1)
X_col = X.columns
X.head()

740       741       742       743       744       745       746  \
0  1.312229  1.313752  1.315363  1.317040  1.318748  1.320468  1.322220   
1  1.207945  1.209409  1.211038  1.212806  1.214665  1.216582  1.218553   
2  1.382931  1.383910  1.385126  1.386556  1.388142  1.389842  1.391645   
3  1.464557  1.465710  1.466986  1.468363  1.469794  1.471253  1.472753   
4  1.145023  1.146660  1.148481  1.150452  1.152523  1.154657  1.156848   

        747       748       749    ...         1061      1062      1063  \
0  1.324037  1.325907  1.327774    ...     1.539160  1.540128  1.539360   
1  1.220593  1.222671  1.224727    ...     1.491851  1.492835  1.492130   
2  1.393558  1.395535  1.397497    ...     1.566846  1.567866  1.567114   
3  1.474325  1.475947  1.477559    ...     1.646721  1.647711  1.646850   
4  1.159107  1.161403  1.163678    ...     1.454652  1.455651  1.454998   

       1064      1065      1066      1067      1068      1069      1070  
0  1.537472  1.535576  1.533732  1.534212  1.536558  1.539456  1.541060  
1  1.490335  1.488528  1.486766  1.487255  1.489549  1.492376  1.493944  
2  1.565218  1.563311  1.561454  1.561961  1.564364  1.567328  1.568971  
3  1.644794  1.642736  1.640737  1.641228  1.643717  1.646801  1.648502  
4  1.453276  1.451538  1.449841  1.450336  1.452587  1.455357  1.456896  

[5 rows x 331 columns]

In [7]:
#collecting depedent variable in Y
Y=df['Predictor']
Y.head()

0    Dimtu
1    Dimtu
2    Dimtu
3    Dimtu
4    Dimtu
Name: Predictor, dtype: object

In [9]:
from sklearn import preprocessing

In [10]:
normalized_X=preprocessing.scale(X)

In [12]:
#Savitzky-Golay filter with second degree derivative.
from scipy.signal import savgol_filter 

sg=savgol_filter(X,window_length=11, polyorder=3, deriv=2, delta=1.0)

C:\Users\LENOVO\Anaconda3\lib\site-packages\scipy\signal\_arraytools.py:45: FutureWarning: Using a non-tuple sequence for multidimensional indexing is deprecated; use `arr[tuple(seq)]` instead of `arr[seq]`. In the future this will be interpreted as an array index, `arr[np.array(seq)]`, which will result either in an error or a different result.
  b = a[a_slice]


In [13]:
sg_x=pd.DataFrame(sg, columns=X_col)

sg_x.head()

740       741       742       743       744       745       746  \
0  0.000069  0.000062  0.000055  0.000048  0.000041  0.000035  0.000022   
1  0.000172  0.000149  0.000127  0.000104  0.000081  0.000059  0.000032   
2  0.000269  0.000236  0.000203  0.000170  0.000137  0.000104  0.000066   
3  0.000112  0.000099  0.000085  0.000072  0.000059  0.000045  0.000026   
4  0.000186  0.000161  0.000137  0.000112  0.000088  0.000063  0.000035   

        747       748       749    ...         1061      1062      1063  \
0  0.000006 -0.000009 -0.000020    ...    -0.000424 -0.000452 -0.000289   
1  0.000005 -0.000019 -0.000036    ...    -0.000416 -0.000442 -0.000284   
2  0.000027 -0.000009 -0.000036    ...    -0.000435 -0.000463 -0.000297   
3  0.000006 -0.000013 -0.000025    ...    -0.000449 -0.000478 -0.000304   
4  0.000007 -0.000018 -0.000035    ...    -0.000411 -0.000437 -0.000282   

           1064      1065      1066      1067      1068      1069      1070  
0  8.533263e-07  0.000294  0.000729  0.001164  0.001599  0.002034  0.002469  
1 -3.527249e-06  0.000280  0.000703  0.001125  0.001547  0.001969  0.002391  
2 -2.368324e-06  0.000296  0.000739  0.001182  0.001625  0.002069  0.002512  
3  5.485790e-06  0.000318  0.000783  0.001248  0.001713  0.002178  0.002643  
4 -7.838702e-06  0.000270  0.000682  0.001094  0.001506  0.001919  0.002331  

[5 rows x 331 columns]

In [14]:
# Splitting the data into train and test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test =train_test_split(sg, Y,
                                                    test_size=0.2,
                                                    random_state=123,stratify=Y)

In [18]:
from sklearn.ensemble import RandomForestClassifier

Rf = RandomForestClassifier(random_state=52)
Rf_fit=Rf.fit(X_train, y_train)

C:\Users\LENOVO\Anaconda3\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


In [19]:
y_pred = Rf.predict(X_test)

In [20]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
print("Test Result:\n")        
print("accuracy score: {0:.4f}\n".format(accuracy_score(y_test, y_pred)))
print("Classification Report: \n {}\n".format(classification_report(y_test, y_pred)))
print("Confusion Matrix: \n {}\n".format(confusion_matrix(y_test,y_pred)))

Test Result:

accuracy score: 0.8474

Classification Report: 
               precision    recall  f1-score   support

  Acos Dubie       0.83      1.00      0.91        10
       Akaki       0.90      0.90      0.90        10
      Ararti       0.80      0.80      0.80        10
       Chefe       1.00      1.00      1.00        10
    DZ-10-11       1.00      0.90      0.95        10
     DZ-10-4       0.91      1.00      0.95        10
      Dalota       1.00      0.90      0.95        10
       Dimtu       0.89      0.80      0.84        10
       Dubie       0.78      0.70      0.74        10
       Ejere       0.50      0.50      0.50        10
       Habru       0.67      0.80      0.73        10
        Hora       0.83      1.00      0.91        10
      Mariye       0.82      0.90      0.86        10
      Minjar       0.80      0.80      0.80        10
      Natoli       1.00      0.90      0.95        10
      Shasho       0.78      0.70      0.74        10
        Teji      

In [22]:
from sklearn.feature_selection import RFE
# RFE with 15 features
rfe_15 = RFE(Rf,15)

# fit with 15 features
rfe_15.fit(X_train, y_train)

# Printing the boolean results
print(rfe_15.support_)

[False False False False False False  True False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False  True False
 False False False False False False False False False False False False
 False False False False False False False  True False False False False
 False False False False False False False False False False False False
 False  True False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False Fa

In [23]:
y_pred = rfe_15.predict(X_test)

print("accuracy score: {0:.4f}\n".format(accuracy_score(y_test, y_pred)))
print("Confusion Matrix: \n {}\n".format(confusion_matrix(y_test, y_pred)))

accuracy score: 0.7421

Confusion Matrix: 
 [[ 8  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  1  1  0]
 [ 0  9  0  0  0  0  0  0  0  0  0  0  1  0  0  0  0  0  0]
 [ 0  0  7  0  0  0  0  0  0  1  0  0  0  0  0  0  0  0  2]
 [ 0  0  0  9  0  0  1  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0 10  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0 10  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  9  0  0  0  1  0  0  0  0  0  0  0  0]
 [ 0  1  0  0  0  0  0  7  0  0  0  0  2  0  0  0  0  0  0]
 [ 0  3  0  0  1  0  0  0  5  0  0  0  0  0  0  0  0  0  1]
 [ 0  0  0  0  0  0  1  0  0  3  2  0  0  1  0  3  0  0  0]
 [ 2  0  0  0  0  0  0  0  0  2  6  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  9  0  0  1  0  0  0  0]
 [ 0  0  0  0  0  0  0  2  0  0  0  0  8  0  0  0  0  0  0]
 [ 0  0  1  0  0  0  0  0  0  1  0  0  0  7  0  0  1  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  1  0  2  7  0  0  0  0]
 [ 0  0  2  0  0  0  0  0  0  1  1  0  0  1  0  3  1  1 

In [24]:
# RFE with 17 features

rfe_17 = RFE(Rf,17)

# fit with 17 features
rfe_17.fit(X_train, y_train)

# Printing the boolean results
print(rfe_17.support_)

[False False False False False False  True False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
  True False False False False False False False False False  True False
 False False False False False False False False False False False False
 False False False False False False False  True False False False False
 False False False False False False False False False False False False
 False  True False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False Fa

In [25]:
y_pred = rfe_17.predict(X_test)

print("accuracy score: {0:.4f}\n".format(accuracy_score(y_test, y_pred)))
print("Confusion Matrix: \n {}\n".format(confusion_matrix(y_test, y_pred)))

accuracy score: 0.7947

Confusion Matrix: 
 [[10  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  9  0  0  0  0  0  0  0  0  0  0  1  0  0  0  0  0  0]
 [ 0  0  9  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  1]
 [ 0  0  0  8  0  0  0  0  0  0  0  0  0  0  0  0  2  0  0]
 [ 0  0  0  0 10  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0 10  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 1  0  0  0  0  0  9  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  9  0  0  0  0  1  0  0  0  0  0  0]
 [ 0  2  0  0  0  0  0  0  6  0  0  0  1  0  0  0  0  0  1]
 [ 0  0  0  0  0  0  0  0  0  2  3  0  0  1  0  3  0  1  0]
 [ 0  0  0  0  0  0  0  0  0  1  9  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  9  0  0  0  0  0  1  0]
 [ 0  0  0  0  0  0  0  1  0  0  0  0  9  0  0  0  0  0  0]
 [ 0  0  1  0  0  0  0  0  0  1  0  0  0  7  0  0  0  1  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  1  0  2  6  0  0  0  1]
 [ 0  0  1  0  0  0  0  0  0  2  0  0  0  0  0  5  1  1 

In [26]:
# RFE with 20 features

rfe_20 = RFE(Rf,20)

# fit with 20 features
rfe_20.fit(X_train, y_train)

# Printing the boolean results
print(rfe_20.support_)

[False False False False False False  True  True False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
  True False False False False False False False False False  True False
 False False False False False False False False False False False False
 False False False False False False False  True False False False False
 False False False False False False False False False False False False
 False  True False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False  True False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False Fa

In [27]:
y_pred = rfe_20.predict(X_test)

print("accuracy score: {0:.4f}\n".format(accuracy_score(y_test, y_pred)))
print("Confusion Matrix: \n {}\n".format(confusion_matrix(y_test, y_pred)))

accuracy score: 0.8000

Confusion Matrix: 
 [[ 9  0  0  0  0  0  0  0  1  0  0  0  0  0  0  0  0  0  0]
 [ 0  7  0  0  0  0  0  0  0  0  0  0  3  0  0  0  0  0  0]
 [ 0  0  9  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  1]
 [ 0  0  0  8  0  0  0  0  0  0  0  0  0  0  0  0  2  0  0]
 [ 0  0  0  0 10  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0 10  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0 10  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  9  0  0  0  0  1  0  0  0  0  0  0]
 [ 0  2  0  0  1  0  0  0  7  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  4  3  0  0  1  0  2  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0 10  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0 10  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  2  0  0  0  0  8  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  1  0  0  0  9  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  2  0  1  7  0  0  0  0]
 [ 0  0  3  0  0  0  0  0  0  5  0  0  0  0  0  2  0  0 

In [28]:
# RFE with 25 features

rfe_25 = RFE(Rf,25)

# fit with 25 features
rfe_25.fit(X_train, y_train)

# Printing the boolean results
print(rfe_25.support_)

[False False False False False False  True  True False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False  True False False False False False False False
  True False False False False False False False False False  True False
 False False False False False False False False False False False False
 False False False False False False False  True False False False False
 False False False False False False False False False False False False
 False  True False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False  True False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False Fa

In [29]:
y_pred = rfe_25.predict(X_test)

print("accuracy score: {0:.4f}\n".format(accuracy_score(y_test, y_pred)))
print("Confusion Matrix: \n {}\n".format(confusion_matrix(y_test, y_pred)))

accuracy score: 0.7737

Confusion Matrix: 
 [[ 9  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  1  0  0]
 [ 0  8  0  0  0  0  0  0  0  0  0  0  2  0  0  0  0  0  0]
 [ 0  0  9  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  1]
 [ 0  0  0  8  0  0  0  0  0  1  0  1  0  0  0  0  0  0  0]
 [ 0  0  0  0 10  0  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0 10  0  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 1  0  0  0  0  0  9  0  0  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  1  0  0  0  7  0  0  0  0  2  0  0  0  0  0  0]
 [ 0  3  0  0  1  0  0  0  6  0  0  0  0  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  2  0  0  2  3  0  0  1  0  2  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  0  9  0  0  0  0  0  0  1  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  7  0  0  2  0  0  1  0]
 [ 0  1  0  0  0  0  0  1  0  0  0  0  8  0  0  0  0  0  0]
 [ 0  0  0  0  0  0  0  0  0  1  0  0  0  8  0  0  0  1  0]
 [ 0  0  0  0  0  0  0  0  0  0  0  0  0  1  9  0  0  0  0]
 [ 0  0  1  0  0  0  0  0  0  2  0  0  0  0  0  5  1  1 

In [30]:
Rf.fit(X_train, y_train)
# Get numerical feature importances
importances = list(Rf.feature_importances_)
# List of tuples with variable and importance
feature_importances = [(feature, round(importance, 4)) for feature, importance in zip(X_col, importances)]
# Sort the feature importances by most important first
feature_importances = sorted(feature_importances, key = lambda x: x[1], reverse = True)
# Print out the feature and importances 
[print('Variable: {:20} Importance: {}'.format(*pair)) for pair in feature_importances]

Variable: 979                  Importance: 0.0153
Variable: 977                  Importance: 0.0138
Variable: 838                  Importance: 0.0134
Variable: 1033                 Importance: 0.0123
Variable: 949                  Importance: 0.0121
Variable: 1015                 Importance: 0.0121
Variable: 1034                 Importance: 0.0117
Variable: 989                  Importance: 0.011
Variable: 991                  Importance: 0.011
Variable: 954                  Importance: 0.0107
Variable: 930                  Importance: 0.0101
Variable: 865                  Importance: 0.0095
Variable: 931                  Importance: 0.009
Variable: 1046                 Importance: 0.009
Variable: 794                  Importance: 0.0088
Variable: 950                  Importance: 0.0087
Variable: 773                  Importance: 0.0085
Variable: 753                  Importance: 0.0084
Variable: 819                  Importance: 0.0083
Variable: 976                  Importance: 0.0083
Vari

[None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,

## Random Forest is giving 84% accuracy and after reducing the variable with RFE the (20 variable ) it is giving as 80% and after furthermore reduction it's decreasing